src/test/java/com/meta/chatbridge/message/FBMessageHandlerTest.java

```diff
@@ -0,0 +1,368 @@
+    int challenge = 1158201444;
+    Pipeline<FBMessage> pipeline =
+        new Pipeline<>(
+            new MemoryStore<>(),
+            new FBMessageHandler("0", token, "dummy"),
+            new DummyFBMessageLLMHandler("this is a dummy message"),
+            "/testfbmessage");
+    final PipelinesRunner runner = PipelinesRunner.newInstance().pipeline(pipeline).port(0);
+    HttpResponse response;
+    try (PipelinesRunner ignored = runner.start()) {
+      ImmutableMap<String, String> params =
+          ImmutableMap.<String, String>builder()
+              .put("hub.mode", "subscribe")
+              .put("hub.challenge", Integer.toString(challenge))
+              .put("hub.verify_token", token)
+              .build();
+      response = getRequest("testfbmessage", runner.port(), params);
+    }
+    assertThat(response.getCode()).isEqualTo(200);
+    InputStream inputStream = ((BasicClassicHttpResponse) response).getEntity().getContent();
+    byte[] input = new byte[inputStream.available()];
+    inputStream.read(input);
+    String text = new String(input, StandardCharsets.UTF_8);
+    assertThat(text).isEqualTo(Integer.toString(challenge));
+  }
+
+  private static Request createMessageRequest(
+      String body, PipelinesRunner runner, boolean calculateHmac)
+      throws IOException, URISyntaxException {
+    @SuppressWarnings("unchecked") // for the scope of this test this is guaranteed
+    Pipeline<FBMessage> pipeline =
+        (Pipeline<FBMessage>) runner.pipelines().stream().findAny().get();
+    String path = pipeline.path();
+
+    // for the scope of this test this is guaranteed
+    FBMessageHandler messageHandler = (FBMessageHandler) pipeline.messageHandler();
+
+    URI uri =
+        URIBuilder.localhost().setScheme("http").setPort(runner.port()).appendPath(path).build();
+
+    Request request = Request.post(uri).bodyString(body, ContentType.APPLICATION_JSON);
+    if (calculateHmac) {
+      String hmac = messageHandler.hmac(body);
+      request.setHeader("X-Hub-Signature-256", "sha256=" + hmac);
+    }
+
+    return request;
+  }
+
+  private static Request createMessageRequest(String body, PipelinesRunner runner)
+      throws IOException, URISyntaxException {
+    return createMessageRequest(body, runner, true);
+  }
+
+  private Function<Identifier, URI> testURLFactoryFactory(Identifier pageId) {
+    return p -> {
+      assertThat(p).isEqualTo(pageId);
+      try {
+        return URIBuilder.localhost().setScheme("http").setPort(app.port()).build();
+      } catch (URISyntaxException | UnknownHostException e) {
+        fail("failed building url");
+        throw new RuntimeException(e);
+      }
+    };
+  }
+
+  private record TestArgument(
+      String name,
+      int expectedReturnCode,
+      ThrowableFunction<PipelinesRunner, Request> requestFactory,
+      boolean messageExpected) {
+
+    // hacky way to try the send twice scenario
+    int timesToSendMessage() {
```

Comment:
Slightly less hacky(?): Have testargument call testargumentwithnumtimes and have that always be 1 and call testargumentwithnumtimes directly in the tests you want to change the argument (or a different interim methiod if TestArgument is needed by the jupiter testing

Reply:
If we make a "defaultTestArg" function it might help for future things we want to set in only one/two tests like this